ci: github action error annotations despite passing tests

Successful test runs cause error annotations in Github Actions, making it easy to miss actual problems with the CI.

E.g. here https://github.com/haskell/cabal/actions/runs/2904212306
<img width="592" alt="Screenshot 2022-08-22 at 16 13 38" src="https://user-images.githubusercontent.com/132113/185942734-0c388939-f1ad-4308-b43d-4b4a270ab035.png">

Raw log output shows this:

```
2022-08-22T13:41:09.8074706Z Main.hs:1:8: error: Variable not in scope: thisNameDoesNotExist
2022-08-22T13:41:09.8075066Z   |
2022-08-22T13:41:09.8075443Z 1 | main = thisNameDoesNotExist
2022-08-22T13:41:09.8075680Z   |        ^^^^^^^^^^^^^^^^^^^^
2022-08-22T13:41:24.6900541Z 
2022-08-22T13:41:24.6928686Z P.hs:4:5: error:
2022-08-22T13:41:24.6957177Z ##[error]    Variable not in scope: this_is_not_expected_to_compile :: Int
2022-08-22T13:41:24.6968223Z   |
2022-08-22T13:41:24.6968691Z 4 | p = this_is_not_expected_to_compile
2022-08-22T13:41:24.6968952Z   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2022-08-22T13:41:27.8968239Z 
2022-08-22T13:41:27.8990740Z P.hs:4:5: error:
2022-08-22T13:41:27.8992090Z ##[error]    Variable not in scope: this_is_not_expected_to_compile :: Int
2022-08-22T13:41:27.8992957Z   |
2022-08-22T13:41:27.8993344Z 4 | p = this_is_not_expected_to_compile
2022-08-22T13:41:27.8993685Z   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2022-08-22T13:41:37.6667032Z 
2022-08-22T13:41:37.6670365Z Q.hs:3:1: error:
2022-08-22T13:41:37.6672015Z ##[error]    Could not load module ‘P’
2022-08-22T13:41:37.6677914Z     It is a member of the hidden package ‘p-0.1’.
2022-08-22T13:41:37.6679288Z     Perhaps you need to add ‘p’ to the build-depends in your .cabal file.
2022-08-22T13:41:37.6680249Z     It is a member of the hidden package ‘p-0.1’.
2022-08-22T13:41:37.6680990Z     Perhaps you need to add ‘p’ to the build-depends in your .cabal file.
2022-08-22T13:41:37.6681451Z     Use -v (or `:set -v` in ghci) to see a list of the files searched for.
2022-08-22T13:41:37.6681684Z   |
2022-08-22T13:41:37.6681864Z 3 | import P
2022-08-22T13:41:37.6682047Z   | ^^^^^^^^
2022-08-22T13:41:49.4512830Z 
2022-08-22T13:41:49.4535576Z All 31 tests passed (129.89s)
2022-08-22T13:41:49.4817071Z [0;32m<<< **/integration-tests2 -j1 --hide-successes --with-ghc=ghc [0m (130/321 sec)
2022-08-22T13:41:49.4817596Z 
2022-08-22T13:41:49.4853596Z ##[group]Run sh validate.sh $FLAGS -s cli-suite
2022-08-22T13:41:49.4854431Z [36;1msh validate.sh $FLAGS -s cli-suite[0m
2022-08-22T13:41:49.4916107Z shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
2022-08-22T13:41:49.4916326Z env:
2022-08-22T13:41:49.4916512Z   GHC_FOR_RELEASE: 9.2.3
2022-08-22T13:41:49.4916825Z   GHC_FOR_SOLVER_BENCHMARKS: 9.2.3
2022-08-22T13:41:49.4917054Z   GHC_FOR_COMPLETE_HACKAGE_TESTS: 9.2.3
2022-08-22T13:41:49.4917273Z   COMMON_FLAGS: -j 2 -v
2022-08-22T13:41:49.4917458Z   FLAGS: -j 2 -v
2022-08-22T13:41:49.4917642Z ##[endgroup]
2022-08-22T13:41:49.5506766Z [0;96m=== cabal-install: cabal-testsuite ============================= 13:41:49 ===[0m
```

The `##[error]` markup seems to be involved. No idea where it comes from, it doesn't seem to be as simple as any `error:` or just stderr output causing this, compare the `thisNameDoesNotExist` line which doesn't cause an annotation.

I don't see ::error, that's the reason of the second part of my message. What I do see is `Variable not in scope: this_is_not_expected_to_compile :: Int` which has a `::`. I thought it may be the reason GitHub adds the ##[error]. But that's a shot in a dark, so feel free to ignore. Also, it doesn't explain why `Could not load module ‘P’` is marked too.

Running the tests locally, I see

```
  Successful builds
    Setup script styles:            OK (21.79s)
      SetupCustomExplicitDeps   (15.85s)
      SetupNonCustomInternalLib (5.94s)
    keep-going:                     
P.hs:4:5: error:
    Variable not in scope: this_is_not_expected_to_compile :: Int
  |
4 | p = this_is_not_expected_to_compile
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

P.hs:4:5: error:
    Variable not in scope: this_is_not_expected_to_compile :: Int
  |
4 | p = this_is_not_expected_to_compile
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
OK (13.84s)
    local tarball:                  OK (11.37s)
  Regression tests
    issue #3324:                    
Q.hs:3:1: error:
    Could not load module ‘P’
    It is a member of the hidden package ‘p-0.1’.
    Perhaps you need to add ‘p’ to the build-depends in your .cabal file.
    It is a member of the hidden package ‘p-0.1’.
    Perhaps you need to add ‘p’ to the build-depends in your .cabal file.
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
  |
3 | import P
  | ^^^^^^^^
OK (18.52s)
    program options scope all:      OK (2.40s)
    program options scope local:    OK (2.32s)
    program options scope specific: OK (2.24s)
```

(with some ANSI colors).

So it *seems* like Github Actions is misinterpreting those lines that end in `error:`. `error::` at the beginning of output is documented here https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message for whatever it's worth.

The best thing would be to suppress that GHC error output entirely within the tests as they look messy and confusing even when run locally.

--- a/cabal-install/cabal-install.cabal
+++ b/cabal-install/cabal-install.cabal
@@ -423,8 +423,10 @@ test-suite integration-tests2
     , directory
     , filepath
     , process
-    , tasty >= 1.2.3 && <1.6
+    , tasty >= 1.5 && <1.6
     , tasty-hunit >= 0.10
+    , tasty-expected-failure
+    , silently
     , tagged
 
 test-suite long-tests

--- a/cabal-install/tests/IntegrationTests2.hs
+++ b/cabal-install/tests/IntegrationTests2.hs
@@ -89,8 +89,13 @@ import System.Process (callProcess)
 
 import Data.Tagged (Tagged (..))
 import Test.Tasty
-import Test.Tasty.HUnit
+import Test.Tasty.ExpectedFailure
+import Test.Tasty.HUnit hiding (testCase)
+import qualified Test.Tasty.HUnit as T (testCase)
 import Test.Tasty.Options
+import Test.Tasty.Runners
+
+import System.IO.Silently
 
 import qualified Data.ByteString as BS
 import Data.Maybe (fromJust)
@@ -112,23 +117,43 @@ main = do
   callProcess "cabal" ["update"]
   defaultMainWithIngredients
     (defaultIngredients ++ [includingOptions projectConfigOptionDescriptions])
-    ( withProjectConfig $ \config ->
-        testGroup
+    ( localOption (NumThreads 1) $ withProjectConfig $ \config ->
+        sequentialTestGroup
           "Integration tests (internal)"
+          AllFinish
           (tests config)
     )
+
+-- Tests are run silently, unless they fail. Firstly because it is annoying to
+-- see lots of stderr from your unit tests. Secondly because this output
+-- leaks into the result of github actions (#8419)
+--
+-- Note that this capture is safe to use as the testsuite runs sequentially.
+silentTest :: TestTree -> TestTree
+silentTest = wrapTest silentHelper
+  where
+    silentHelper t = do
+      (out, res) <- hCapture [stderr] t
+
+      return $
+        if not (resultSuccessful res)
+          then res{resultDescription = resultDescription res <> "\nCaptured output:\n" <> out}
+          else res
+
+testCase :: String -> Assertion -> TestTree
+testCase desc action = (T.testCase desc action)
 
 tests :: ProjectConfig -> [TestTree]
 tests config =
   -- TODO: tests for:
   -- \* normal success
   -- \* dry-run tests with changes
-  [ testGroup "Discovery and planning" $
+  [ sequentialTestGroup "Discovery and planning" AllFinish $
       [ testCase "no package" (testExceptionInFindingPackage config)
       , testCase "no package2" (testExceptionInFindingPackage2 config)
       , testCase "proj conf1" (testExceptionInProjectConfig config)
       ]
-  , testGroup "Target selectors" $
+  , sequentialTestGroup "Target selectors" AllFinish $
       [ testCaseSteps "valid" testTargetSelectors
       , testCase "bad syntax" testTargetSelectorBadSyntax
       , testCaseSteps "ambiguous syntax" testTargetSelectorAmbiguous
@@ -145,7 +170,7 @@ tests config =
       , testCaseSteps "problems (bench)" (testTargetProblemsBench config)
       , testCaseSteps "problems (haddock)" (testTargetProblemsHaddock config)
       ]
-  , testGroup "Exceptions during building (local inplace)" $
+  , sequentialTestGroup "Exceptions during building (local inplace)" AllFinish $
       [ testCase "configure" (testExceptionInConfigureStep config)
       , testCase "build" (testExceptionInBuildStep config)
       --    , testCase "register"   testExceptionInRegisterStep
@@ -154,7 +179,7 @@ tests config =
     -- TODO: need to check we can build sub-libs, foreign libs and exes
     -- components for non-local packages / packages in the store.
 
-    testGroup "Successful builds" $
+    sequentialTestGroup "Successful builds" AllFinish $
       [ testCaseSteps "Setup script styles" (testSetupScriptStyles config)
       , testCase "keep-going" (testBuildKeepGoing config)
       ]
@@ -164,19 +189,20 @@ tests config =
           else
             [ testCase "local tarball" (testBuildLocalTarball config)
             ]
-  , testGroup "Regression tests" $
+  , sequentialTestGroup "Regression tests" AllFinish $
       [ testCase "issue #3324" (testRegressionIssue3324 config)
       , testCase "program options scope all" (testProgramOptionsAll config)
       , testCase "program options scope local" (testProgramOptionsLocal config)
       , testCase "program options scope specific" (testProgramOptionsSpecific config)
       ]
-  , testGroup "Flag tests" $
+  , sequentialTestGroup "Flag tests" AllFinish $
       [ testCase "Test Nix Flag" testNixFlags
       , testCase "Test Config options for commented options" testConfigOptionComments
       , testCase "Test Ignore Project Flag" testIgnoreProjectFlag
       ]
-  , testGroup
+  , sequentialTestGroup
       "haddock-project"
+      AllFinish
       [ testCase "dependencies" (testHaddockProjectDependencies config)
       ]
   ]

